### DISCUSSION TITLE
Ability to host your own WrapDB

### DISCUSSION THREAD

**Discussion by LunarLambda** [mesonbuild/meson#12296] [Ideas]
I maintain a [large list of meson wraps for Game Boy Advance homebrew projects](https://github.com/LunarLambda/meson-gba/tree/main/subprojects).

Currently, making these available to users requires they download the meson-gba skeleton project, which includes all files for all libraries and tools unconditionally. This is both messy for the end user, and messy for me to maintain and test (currently I have a `dev` branch I need to keep somewhat painstakingly rebasing against master, since my development setup and the end user setup are different). It also means to get an updated version of a wrap, the end user needs to either manually download the files, or do a `git stash; git pull; git stash apply` dance)

Meson provides `meson wrap` for downloading and adding wraps from a hosted database. However, there's no mechanism to use a database other than the meson-provided WrapDB. Besides being a potential single point of failure (no ability to use mirror servers), this means I am left with two options, either reimplementing a form of WrapDB for my project, or continue doing it this way. The highly domain-specific nature of these projects means they're unlikely candidates for the official WrapDB, and even without that, the overhead of making semi-frequent pull requests to WrapDB for on the order of two dozen projects is simply not feasible for me.

It would be great if meson could allow specifying a WrapDB server, or a similar functionality, such as being able to download a wrap from a WrapDB-*like* git repository, for example.

**Comment by LunarLambda** [comment]
https://github.com/mesonbuild/meson/issues/6497#issuecomment-578245322 Apparently the service backing WrapDB is already open source and could (possibly?) be used for hosting, so it's just a matter of allowing to specify a source for `meson wrap` to use.

**Reply by xclaesse** [reply]
That's the old server implementation we are not using anymore. WrapDB v2 is much simpler, all it needs is a static URL to a releases.json file.

**Comment by xclaesse** [comment]
3rd party WrapDB has always been something we wanted to allow. It's not currently possible, but it's shouldn't be too hard to get there.

There is only a handful places in Meson where the wrapdb URL is hardcoded:
https://github.com/mesonbuild/meson/blob/master/mesonbuild/wrap/wrap.py#L111

Basically what would be needed is a command line argument to "meson wrap" to specify a custom URL where to find a releases.json file.

**Reply by LunarLambda** [reply]
That's great to hear. Is something like this in progress already, or is there a reason it hasn't been done yet? I may have time to look into making a PR for it, it would genuinely be a killer feature for me.

**Reply by eli-schwartz** [reply]
I am not sure what "always wanted to allow" means other than a personal desire of @xclaesse.

From my own perspective it's something that now that it has been mentioned, I have no objection to. At the same time, it's just not something I'm personally interested in, so I wouldn't write a patch for it or put it on my priority list, but I would accept a patch if someone else wrote one.

**Reply by LunarLambda** [reply]
What would an acceptable design for that be? would a `--url` option for `meson wrap` be sufficient?

**Reply by xclaesse** [reply]
Maybe `--wrapdb-url`. But you'll have to add releases.json in your git repository, and setup your CI the same way as we do in WrapDB to create release tags, probably copying our script: https://github.com/mesonbuild/wrapdb/blob/master/tools/create_release.py. We might have to think about an easier solution tbh.

**Reply by LunarLambda** [reply]
gotcha.

Also, I think meson wrap downloads some .json file to the subprojects directory, right?

Could I theoretically just handwrite/generate one and ship that in the meantime? I didn't have time to investigate it, I just randomly spotted it while reading the docs. If it's good enough to let me hoist the packagefiles elsewhere, that's fine enough

**Reply by xclaesse** [reply]
`meson wrap update-db` writes releases.json into subprojects/wrapdb.json. That's a way for not having to do `meson wrap install` each wrap file individually. We should add into that json file the URL of the wrapdb instance you're using. But you'll still need to generate release tags that contains the patch overlay tarball. That's what the script I linked above does. One could of course write a server that would generate those tarballs on the flight I guess.

**Reply by LunarLambda** [reply]
So a patch for this would be twofold, one to add the CLI argument and one to add the URL field to releases/wrapdb.json?

**Reply by xclaesse** [reply]
I think when `update-db` subcommand writes `subprojects/wrapdb.json`, it whould add a field in that json with the URL from where it downloaded it. That way, in wrap.py get_from_wrapdb() you can fetch the .wrap file from the right URL.

**Reply by xclaesse** [reply]
BTW, all those https://wrapdb.mesonbuild.com/v2 URLs work with a simple server we have that redirect to github: https://github.com/mesonbuild/wrapdb/blob/master/nginx/default.

**Comment by dreamer-coding** [comment]
I was hoping to see an answer to this question, I've been asking the same question with no response.

**Reply by eli-schwartz** [reply]
It is not a question. Therefore it will never be answered.

It is a proposal. Nobody has attempted to implement it yet.

**Reply by dreamer-coding** [reply]
Why not? Are there any plans to get around to this possible feature?

**Reply by eli-schwartz** [reply]
Most commonly for FOSS projects, feature requests (that is, not regression bugs) have no set priorities and therefore only get implemented whenever someone has enough personal interest in the feature to devote time and attention to it.

Look up earlier in this discussion. I already said two years ago: I am not personally interested in the feature and I will NOT be implementing it myself, but I will review patches if someone else is interested in submitting a Pull Request for it.

It is possible that there is another Meson maintainer who is personally interested in the feature and will implement it, but it's been two years and nobody has done so. Therefore my advice is to not assume someone else will do it -- if you're interested in the feature, submit a PR for it yourself, and take me up on my offer to review it.

I am sorry, but that is the best offer I can provide.

**Reply by dreamer-coding** [reply]
Well I nominate myself to attempt this and to add Kotlin to Meson build, provide pull request for this winter, like to get these two features right.
